### PR TITLE
Resource property validation should be aware of action-specific reqs

### PR DESCRIPTION
**Pull request by sabat**
## Description

The `get` and `set` methods in the Property class, meant to be used by resources, validate that a property is required if its value is nil. However, they don't check to see whether the property was required for the *specific action* the resource is using, so any use of `get` and `set` will throw an exception if used with a resource that doesn't use a required property, even when the property is not required for that action.

That's a vague explanation, so I'll get specific: the current version of the chef-client cookbook has a recipe (cron.rb) with a resource that does a delete. When you try to get that resource, it throws an exception, explaining that `:command` is a required property—but it's only required for the `:create` action, not for `:delete`. https://github.com/chef-cookbooks/chef-client/blob/master/recipes/cron.rb#L94-L96

This PR simply fixes that by passing the resource's action to the `required?` method, and allows the `required?` method to compare arrays of actions rather than just expecting a single action.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [X] I have read the **CONTRIBUTING** document.
- [X] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [X] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

**Comment by lamont-granquist** [comment]
Didn't I already fix this in #9688?

**Comment by sabat** [comment]
> Didn't I already fix this in #9688?

I don't know, but the bug absolutely exists. Property#get and Property#set do not check whether a property is required by a specific action, so when Property#required? is called, it's without an argument. That leads to undesirable situations like the chef-client cookbook's cron recipe failing introspection. (Details of that are in the description.)

TL;DR: not sure what you fixed in #9688 but there's still a problem.

**Comment by lamont-granquist** [comment]
Oh yeah, I've seen that bug before, it just fell off my radar

Your merge commits have created a bit of of a mess in this PR though.  Might want to just take the diffs you need and start over fresh.

**Comment by sabat** [comment]
Yeah, I was trying to stay up to date with master and wound up with a mess. I'll close this one and resubmit.
